hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -1558,7 +1573,10 @@ def block_len(i):
         def reoffset_pair(pair, o):
             n = len(self.blocks)
             # Number of blocks may have changed, need to validate
-            valid_pair = [p for p in pair if p < n and int_from_block(p) > 0]
+            valid_pair = [
+                p for p in pair if p < n and int_from_block(p) > 0 and
+                self.is_payload_block(p)
```

Comment:
nit: the condition is split across two lines, right? I wonder if rewriting to put the condition on one line would be more readable:

```python
valid_pair = [
    p
    for p in pair
    if p < n and int_from_block(p) > 0 and self.is_payload_block(p)
]
```